### ISSUE TITLE
[CT-2455] [Feature] Less strict secret env vars: allow elsewhere besides just `profiles.yml` + `packages.yml`

### ISSUE THREAD

**Issue by jeremyyeo** [dbt-labs/dbt-core#7425]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

I think this one just calls for a reversal of the decision made in https://github.com/dbt-labs/dbt-core/issues/4310

One example where you might want to use secret env vars is when using Snowflake's [`encrypt_x / decrypt_x` functions](https://docs.snowflake.com/en/sql-reference/functions-encryption)... e.g.

```sql
-- models/pii.sql
select encrypt('john', '{{ env_var("DBT_ENV_SECRET_PWD") }}') as name
```

Which due to the issue/pr above results in the expected error:

```
Parsing Error
  Secret env vars are allowed only in profiles.yml or packages.yml. Found 'DBT_ENV_SECRET_PWD' referenced elsewhere.
```

### Describe alternatives you've considered

I don't think any good alternatives exist.

1. You use a normal var - which exposes the passphrase in plaintext.
2. On snowflake - you could store the passphrase in a table - setup datamasking/role based access and what not - kinda convoluted.

### Who will this benefit?

Anyone wanting to use secret vars in places that are not just those 2 files.

### Are you interested in contributing this feature?

Sure

**Comment by jtcohen6** [comment]
@jeremyyeo There's a workaround for this today:
- Create two env vars, one with the "secret" prefix and one without (e.g. `DBT_ENV_SECRET_PWD` and `SECRET_PWD`) set to the same value
- Use the "non-secret" env var inside the SQL file

```
-- models/pii.sql
select encrypt('john', '{{ env_var("SECRET_PWD") }}') as name
```

The value will be scrubbed from dbt's logs:
```
$ SECRET_PWD=abc123 DBT_ENV_SECRET_PWD=abc123 dbt compile -s pii
12:20:52  Running with dbt=1.5.0-rc2
...
12:20:55  Compiled node 'pii' is:
-- models/pii.sql
select encrypt('john', '*****') as name
```
```sql
create or replace   view analytics.dbt_jcohen.pii
  
   as (
    -- models/pii.sql
select encrypt('john', '*****') as name
  );
```

And _for a completely separate reason_, they will also be scrubbed from Snowflake's query logs, because of the nature of the function (`encrypt`) being used:

<img width="1312" alt="Screenshot 2023-04-21 at 14 07 56" src="https://user-images.githubusercontent.com/13897643/233634063-0ad86d02-9775-4975-b0dc-36ae56b4859e.png">
<img width="658" alt="Screenshot 2023-04-21 at 14 16 06" src="https://user-images.githubusercontent.com/13897643/233634066-2da9d7dc-8fd8-4392-b793-4c5b6a60bb73.png">

The downside of this approach is, _by also storing the value in a non-secret env var_, there's nothing preventing someone with development access to the same project from writing that value to a database table (outside a function like `encrypt` that Snowflake knows to scrub), and then querying the table to read it back out. That's _not_ possible with a "secret" env var due to the intentional limitations around where it can be used/rendered.

Longer term, we should really reconsider how we treat secrets (#6353). Currently, it's just a special behavior for scrubbing a value from the dbt (logs) if it matches an environment variable with a certain prefix.

**Comment by jeremyyeo** [comment]
I guess this may be one that could be more dbt Cloud related? But is directly the result of dbt-core not supporting secrets in other places (one of those "cross-functional" thingies =]). Why would one use secret env vars in the first place?
a) You don't want it logged in logs.
b) You don't want all dbt developers to know the secret value in the first place.

I think what you mentioned addresses (a) but not (b)?

**Comment by jeremyyeo** [comment]
I wrote this elsewhere but worth putting down here too - recapping why simply putting env vars in model files doesn't solve the problem at hand.

Let's assume we want to use the encrypt function (for simplicity).
 
And assuming a toy example:
 
```sql
-- some_model.sql
select encrypt('john', 'correct horse battery staple') as name
```

Okay, so when we build this table in Snowflake - what DDL/DML does dbt send? First things first... dbt has to compile the above model into working and proper SQL:

```sql
-- target/compiled/some_model.sql
create table some_model as 
select encrypt('john', 'correct horse battery staple') as name;
```

Now... let's assume for a second that dbt can have secret env vars in model files:

```sql
-- some_model.sql
select encrypt('john', '{{ env_var("DBT_ENV_SECRET_PASSWORD") }}') as name
```

Now... what happens?
 
Firstly, dbt will always have to compile the sql model file and output it to the target folder - just like above. So - if dbt were to MASK the values in the compiled SQL:
 
```sql
-- target/compiled/some_model.sql
create table some_model as 
select encrypt('john', '!!!!*****!!!!') as name;
 ```

^ Then you table some_model will literally be using the string `'!!!!*****!!!!'` as the encryption passphrase - as opposed to using the env var set by DBT_ENV_SECRET_PASSWORD (which would have been equal to the string `'correct horse battery staple'`).

Because dbt always always has to compile to valid SQL first - that's kind of the rationale that secret env vars were explicitly blocked from model files.

----

Separately I went about testing out using Snowflake Data Masking for this.

![image](https://user-images.githubusercontent.com/8496748/235004758-520c767f-d433-4309-b624-e63d68238eb8.png)

I created a secret table - whose value is only visible to `TRANSFORMER` and not` DEVELOPER` - I then tried to use the `DEVELOPER` role to try and encrypt it's value.

![image (1)](https://user-images.githubusercontent.com/8496748/235004830-cbbe04db-9d56-46c6-a3ac-6b42beb75fcd.png)

It failed with the passphrase 'correct horse battery staple'.

![image (2)](https://user-images.githubusercontent.com/8496748/235004883-6f084f27-4285-4269-a596-800f89656365.png)

But succeeded with the passphrase that was 5 asterisk - the masked value. Meaning it had been encrypted not with 'correct horse battery staple' but '*****' - so unfortunately this looks like a dead end as well unfortunately :(

**Comment by jeremyyeo** [comment]
Reopening on behalf of an enterprise customer :)

**Comment by RobinSmithTHR** [comment]
I have to agree w/the original poster, limiting secret env vars to profiles.yml and packages.yml is ok w/Core, but falls very short when using cloud as profiles.yml is not an option.  Given the differences between environments, within Cloud, accessing the ENV variables directly should be appropriate.  Without this functionality, the use of python scripts which need credentials is a non-starter.

Please consider a correction here for Cloud, otherwise any python+api integrations are unusable.

**Comment by abhra-gupta-trakstar** [comment]
> I have to agree w/the original poster, limiting secret env vars to profiles.yml and packages.yml is ok w/Core, but falls very short when using cloud as profiles.yml is not an option. Given the differences between environments, within Cloud, accessing the ENV variables directly should be appropriate. Without this functionality, the use of python scripts which need credentials is a non-starter.
> 
> Please consider a correction here for Cloud, otherwise any python+api integrations are unusable.

We are facing the same issue in DBT CLOUD Enterprise in being unable to use Encryption Keys as `DBT_ENV_SECRET_` prefixed variables.
We need to be able to use our encryption key in the most secured way possible ideally masked. Having the encryption key exposed like a string under Env Variables list makes us nervous. This encryption key is used to encrypt sensitive client information. 
We want to be able to use this variable the same way as we use Airflow Variables. When we name the variables in a certain way, the variables are masked in the Variables list.

**Comment by manju-choudhary** [comment]
Adding my perspective here - In dbt Cloud Enterprise, the lack of support for securely handling secrets (like encryption keys) creates a major gap. Right now, sensitive values show up as plain strings in the environment variable list, which is not ideal when dealing with client data encryption.

What we really need is a way to manage secrets that aligns with how other platforms (like Airflow) mask variables automatically when named in specific ways. This would let us safely use keys and credentials inside Python or API integrations without exposing them directly. Until there’s a secure masking mechanism for environment variables in cloud, using encryption keys and other credentials feels too risky.

Are there any plans to support masked variables like DBT_ENV_SECRET_ in the future?
